File: craft/commands/rdns.py
import click

from craft.client import delete, get, put
from craft.output import print_item, print_success


@click.command("get")
@click.argument("vm_id", required=False, default=None)
def rdns_get(vm_id):
    """Get reverse DNS record."""
    if not vm_id:
        from craft.interactive import select_vm
        vm_id = select_vm()
    data = get(f"/vms/{vm_id}/rdns")
    print_item(data)


@click.command("set")
@click.argument("vm_id", required=False, default=None)
@click.option("--hostname", default=None, help="FQDN (max 253 chars)")
def rdns_set(vm_id, hostname):
    """Set reverse DNS hostname."""
    if not vm_id:
        from craft.interactive import select_vm
        vm_id = select_vm()
    if not hostname:
        from craft.interactive import input_text
        hostname = input_text("Hostname (FQDN):")
    put(f"/vms/{vm_id}/rdns", {"hostname": hostname})
    print_success(f"rDNS set to {hostname}")


@click.command("delete")
@click.argument("vm_id", required=False, default=None)
def rdns_delete(vm_id):
    """Remove reverse DNS record."""
    if not vm_id:
        from craft.interactive import select_vm
        vm_id = select_vm()
    delete(f"/vms/{vm_id}/rdns")
    print_success("rDNS record removed.")

File: cmd/rdns.go
package cmd

import (
	"fmt"

	"github.com/CraftIntertech/craft-cli/internal/client"
	"github.com/CraftIntertech/craft-cli/internal/output"
	"github.com/spf13/cobra"
)

var rdnsGetCmd = &cobra.Command{
	Use:   "get [VM_ID]",
	Short: "Get reverse DNS record for a VM",
	Run: func(cmd *cobra.Command, args []string) {
		vmID := ""
		if len(args) > 0 {
			vmID = args[0]
		}
		if vmID == "" {
			vmID = SelectVM("Select VM for rDNS")
		}

		result := client.Get(fmt.Sprintf("/vms/%s/rdns", vmID), nil, true)
		output.PrintItem(extractObj(result))
	},
}

var rdnsSetCmd = &cobra.Command{
	Use:   "set [VM_ID]",
	Short: "Set reverse DNS record for a VM",
	Run: func(cmd *cobra.Command, args []string) {
		vmID := ""
		if len(args) > 0 {
			vmID = args[0]
		}
		if vmID == "" {
			vmID = SelectVM("Select VM for rDNS")
		}

		hostname, _ := cmd.Flags().GetString("hostname")
		if hostname == "" {
			hostname = PromptText("Enter FQDN hostname:")
		}

		body := map[string]interface{}{
			"hostname": hostname,
		}

		result := client.Put(fmt.Sprintf("/vms/%s/rdns", vmID), body, true)
		output.PrintSuccess("Reverse DNS record set.")
		output.PrintItem(extractObj(result))
	},
}

var rdnsDeleteCmd = &cobra.Command{
	Use:   "delete [VM_ID]",
	Short: "Delete reverse DNS record for a VM",
	Run: func(cmd *cobra.Command, args []string) {
		vmID := ""
		if len(args) > 0 {
			vmID = args[0]
		}
		if vmID == "" {
			vmID = SelectVM("Select VM for rDNS deletion")
		}

		if !Confirm("Are you sure you want to delete the reverse DNS record?") {
			output.PrintInfo("Cancelled.")
			return
		}

		client.Delete(fmt.Sprintf("/vms/%s/rdns", vmID), true)
		output.PrintSuccess("Reverse DNS record deleted.")
	},
}

func init() {
	rdnsSetCmd.Flags().String("hostname", "", "FQDN hostname for reverse DNS")

	RDNSCmd.AddCommand(rdnsGetCmd)
	RDNSCmd.AddCommand(rdnsSetCmd)
	RDNSCmd.AddCommand(rdnsDeleteCmd)
}
